src/server/node_manager.c

```diff
@@ -6850,124 +6895,69 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
 
-	DBPRT(("%s: entered\n", __func__))
+	struct pbsnode *pnode;
+	mom_svrinfo_t *psvrmom;
+	char *execvnod_in = NULL;
+	char *execvncopy;
+	char *chunk;
+	char *last;
+	int hasprn;
+	char *vname;
+	int nelem;
+	struct key_value_pair *pkvp;
+	char *execvnod = NULL;
 
-	/* decrement number of jobs on the Mom who is the first Mom */
-	/* for the job, Mother Superior; incremented in set_nodes() */
-	/* and saved in ji_destin in assign_hosts()		    */
-	if (((pjob->ji_qs.ji_svrflags & JOB_SVFLG_HasNodes) != 0) &&
-		(pjob->ji_qs.ji_destin[0] != '\0')) {
-		pnode = find_nodebyname(pjob->ji_qs.ji_destin);
-		if (pnode) {
-			psvrmom = pnode->nd_moms[0]->mi_data;
-			if (--psvrmom->msr_numjobs < 0)
-				psvrmom->msr_numjobs = 0;
-		}
+	
+	/* decrement number of jobs on the Mom who is the first Mom */	
+	/* for the job, Mother Superior; incremented in set_nodes() */	
+	/* and saved in ji_destin in assign_hosts()		    */	
+	if (((pjob->ji_qs.ji_svrflags & JOB_SVFLG_HasNodes) != 0) &&	
+		(pjob->ji_qs.ji_destin[0] != '\0')) {	
+		pnode = find_nodebyname(pjob->ji_qs.ji_destin);	
+		if (pnode) {	
+			psvrmom = pnode->nd_moms[0]->mi_data;	
+			if (--psvrmom->msr_numjobs < 0)	
+				psvrmom->msr_numjobs = 0;	
+		}	
 	}
-
+	
 	/* Now loop through the Moms and remove the jobindx entry */
-	/* and remove this jobs's jobinfo entry from each vnode   */
-	for (i = 0; i < mominfo_array_size; i++) {
-		if (mominfo_array[i] == NULL)
-			continue;
-		psvrmom = (mom_svrinfo_t *)(mominfo_array[i]->mi_data);
-
-		for (j=0; j<psvrmom->msr_jbinxsz; j++) {
-			if (psvrmom->msr_jobindx[j] == pjob) {
-				psvrmom->msr_jobindx[j] = NULL;
-				if (is_called_by_job_purge)
-					special_case = 1;
-			}
-		}
-
-		if (special_case) {
-			snprintf(log_buffer, LOG_BUF_SIZE, "\n================================================================"
-				"======================================================\nPBS diagnostic information."
-				" Share this log with PBS Team.\n=========================================="
-				"============================================================================\n"
-				"Mom's state:%lu, number of jobs on this node: %d, number of vnodes: %d.\n"
-				"Other jobs present in the node follows:",
-				psvrmom->msr_state, psvrmom->msr_numjobs, psvrmom->msr_numvnds);
-			log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB, LOG_INFO, pjob->ji_qs.ji_jobid, log_buffer);
-			for (j=0; j<psvrmom->msr_jbinxsz; j++)
-				if (psvrmom->msr_jobindx[j] != NULL)
-					log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB, LOG_INFO, pjob->ji_qs.ji_jobid, psvrmom->msr_jobindx[j]->ji_qs.ji_jobid);
-			sprintf(log_buffer, "===================================================================="
-				"==================================================");
-			log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB, LOG_INFO, pjob->ji_qs.ji_jobid, log_buffer);
-		}
+	/*  remove this jobs's jobinfo entry from each vnode   */
+	
+	if (pjob->ji_wattr[(int)JOB_ATR_exec_vnode].at_flags & ATR_VFLAG_SET) {
```

Comment:
Do we really need this if/else? old code was not checking for this, and I don't think free_nodes() will be called if job is allocated any nodes

Reply:
It is a good practice to check the attribute flag and retrieve it. I am seeing this practice everywhere. please refer same file, you could see many references.

Reply:
I agree that its good practice checks things before accessing, but we can improve this if/else as below:
```c
if ((pjob->ji_wattr[JOB_ATR_exec_vnode].at_flags & ATR_VFLAG_SET) == 0 || ((execvnod_in = pjob->ji_wattr[JOB_ATR_exec_vnode].at_val.at_str) == NULL)) {
	log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_JOB, LOG_DEBUG, pjob->ji_qs.ji_jobid, "in free_nodes and no exec_vnode");
	return;
}
```

Reply:
I have changed this code.